validation/url_checker.py

```diff
@@ -0,0 +1,227 @@
+
+  Returns:
+    status: Status code and reason.
+    message: Optional message providing additional details about response.
+  """
+  try:
+    resp = requests.get(url, timeout=30)
+
+    # Redirect.
+    if resp.history:
+      initial_resp = resp.history[0]
+      return str(initial_resp.status_code
+                ) + " " + initial_resp.reason if initial_resp.reason else str(
+                    initial_resp.status_code), "Redirected to " + resp.url
+
+    return str(resp.status_code) + " " + resp.reason if resp.reason else str(
+        resp.status_code), ""
+
+  except Exception as e:
+    return "ERROR", str(e)
+
+
+def get_modified_files():
+  """Returns list of locally modified MCF files in the current branch.
+
+  Returns:
+    List of modified files.
+  """
+  repo = git.Repo('.', search_parent_directories=True)
+  return [
+      "../" + file for file in sorted(
+          filter(
+              lambda x: x.endswith(".mcf"),
+              list(item.a_path for item in repo.index.diff(None)) +
+              list(repo.untracked_files)))
+  ]
+
+
+def find_invalid_manifest_urls(allowlisted_urls):
+  """Finds invalid urls from manifest using the DC API.
+
+  Args:
+    allowlisted_urls: Set of urls to skip checking.
+
+  Returns:
+    Map of ("MANIFEST", <url>) -> (<status>, <message>) for invalid urls.
+  """
+  dc_api_key = os.environ.get('DC_API_KEY', '')
+
+  resp = requests.get(
+      API_PREFIX + dc_api_key +
+      "&nodes=Provenance&nodes=Dataset&nodes=Source&property=<-typeOf")
+  if resp.status_code != 200:
+    raise ValueError(API_ERROR.format(resp.status_code))
+
+  dcids = []
+  for _, data in resp.json()["data"].items():
+    for node in data["arcs"]["typeOf"]["nodes"]:
+      dcids.append(node["dcid"])
+
+  resp = requests.get(API_PREFIX + dc_api_key +
+                      "".join(["&nodes=" + dcid for dcid in dcids]) +
+                      "&property=->[" + ",".join(URL_PROPS) + "]")
+  if resp.status_code != 200:
+    raise ValueError(API_ERROR.format(resp.status_code))
+
+  urls = set()
+  for _, data in resp.json()["data"].items():
+    if "arcs" not in data:
+      continue
+    for _, nodes in data["arcs"].items():
+      for node in nodes["nodes"]:
+        if (MANIFEST, node["value"]) not in allowlisted_urls:
+          urls.add(node["value"])
+
+  invalid_urls = {}
+  for url in sorted(urls):
+    print("Checking", url)
+    status, message = get_status(url)
+    if not status.startswith("200"):
+      invalid_urls[(MANIFEST, url)] = (status, message)
+
+
+def find_invalid_schema_urls(allowlisted_urls, files):
+  """Finds invalid urls from schema files using the DC API.
+
+  Args:
+    allowlisted_urls: Set of urls to skip checking.
+
+  Returns:
+    Map of (<file>, <url>) -> (<status>, <message>) for invalid urls.
+  """
+  invalid_urls = {}
+  all_urls = set()
+  for file in files:
+    with open(file) as f_in:
+      print("\n*** Processing", file, "***\n")
+      text = f_in.read()
```

Comment:
Wonder if we should restrict to certain properties (description, name, url, descriptionUrl, etc) ?

Reply:
That's true, I think right now the bottleneck is "footnote" from the SDG import. I can look into this as a follow up